src/daft-scan/src/scan_task_iters.rs

```diff
@@ -177,137 +178,225 @@ impl<'a> Iterator for MergeByFileSize<'a> {
+        let curr_row_group_indices = std::mem::take(&mut self.row_group_indices);
+        let curr_size_bytes = std::mem::take(&mut self.size_bytes);
+        let curr_num_rows = std::mem::take(&mut self.num_rows);
+
+        // Create a new DataSource by mutating a clone of the old one
+        let mut new_source = ctx.data_source().clone();
+        if let DataSource::File {
+            chunk_spec,
+            size_bytes,
+            parquet_metadata,
+            metadata: table_metadata,
+            ..
+        } = &mut new_source
+        {
+            // Create a new Parquet FileMetaData, only keeping the relevant row groups
+            let row_group_list = RowGroupList::from_iter(
+                curr_row_group_indices
+                    .iter()
+                    .map(|idx| (*idx, ctx.file_metadata.row_groups.get(idx).unwrap().clone())),
+            );
+            let new_metadata = ctx
+                .file_metadata
+                .clone_with_row_groups(curr_num_rows, row_group_list);
+            *parquet_metadata = Some(Arc::new(new_metadata));
+
+            // Mutate other necessary metadata
+            *chunk_spec = Some(ChunkSpec::Parquet(
+                curr_row_group_indices
+                    .into_iter()
+                    .map(|idx| idx as i64)
+                    .collect(),
+            ));
+            *size_bytes = Some(curr_size_bytes as u64);
+            if let Some(metadata) = table_metadata {
+                metadata.length = curr_num_rows;
+            }
+        } else {
+            unreachable!("Parquet file format should only be used with DataSource::File");
+        }
+
+        Some(new_source)
+    }
+}
+
+struct SplitParquetByRowGroupsContext<'a> {
+    config: &'a DaftExecutionConfig,
+    file_metadata: FileMetaData,
+    scan_task_ref: ScanTaskRef,
+    scantask_estimated_size_bytes: Option<usize>,
+}
 
-                        let mut file = io_runtime.block_on_current_thread(read_parquet_metadata(
+impl<'a> SplitParquetByRowGroupsContext<'a> {
+    pub fn new(
+        scan_task_ref: ScanTaskRef,
+        file_metadata: FileMetaData,
+        config: &'a DaftExecutionConfig,
+    ) -> Self {
+        let scantask_estimated_size_bytes =
+            scan_task_ref.estimate_in_memory_size_bytes(Some(config));
+        Self {
+            config,
+            file_metadata,
+            scan_task_ref,
+            scantask_estimated_size_bytes,
+        }
+    }
+
+    #[inline]
+    fn total_rg_compressed_size(&self) -> usize {
+        self.file_metadata
+            .row_groups
+            .iter()
+            .map(|rg| rg.1.compressed_size())
+            .sum()
+    }
+
+    #[inline]
+    fn data_source(&self) -> &DataSource {
+        match self.scan_task_ref.sources.as_slice() {
+            [source] => source,
+            _ => unreachable!(
+                "SplitByRowGroupsAccumulator should only have one DataSource in its ScanTask"
+            ),
```

Comment:
This seems like a fairly odd constraint that I inherited from the previous logic. I think we should be able to also split ScanTasks with more than 1 DataSource, but not a priority atm.